File: fsharp/collatz-conjecture/CollatzConjecture.fs
module CollatzConjecture

let rec step (n: int, stepcount: int): int =
    if n = 1 then
        stepcount
    else
        let newnum = if n % 2 = 0 then n / 2 else 3 * n + 1
        step(newnum, stepcount + 1)
    
let steps (number: int): int option =
    if number < 1 then
        None
    else
        let count = step(number, 0)
        Some count

File: scala/collatz-conjecture/src/main/scala/CollatzConjecture.scala
object CollatzConjecture {
    def steps(n: Int): Option[Int] = {
        if (n > 0) Some(step(n, 0))
        else None
    }

    def step (n: Int, stepcount: Int): Int = {
        if (n == 1) 
            stepcount
        else {
            val newnum: Int = if (n % 2 == 0) n / 2 else 3 * n + 1
            step(newnum, stepcount + 1)
        }
    }
}